pkg/cmds/backup_mongo.go

```diff
@@ -0,0 +1,159 @@
+package cmds
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/appscode/go/flags"
+	"github.com/appscode/stash/pkg/restic"
+	"github.com/spf13/cobra"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+	appcatalog_cs "kmodules.xyz/custom-resources/client/clientset/versioned"
+)
+
+const (
+	JobMongoBackup  = "stash-mongo-backup"
+	MongoUser       = "username"
+	MongoPassword   = "password"
+	MongoDumpFile   = "dumpfile.sql"
+	MongoDumpCMD    = "mongodump"
+	MongoRestoreCMD = "mongorestore"
+)
+
+func NewCmdBackupMongo() *cobra.Command {
+	var (
+		masterURL      string
+		kubeconfigPath string
+		namespace      string
+		appBindingName string
+		mongoArgs      string
+		outputDir      string
+		setupOpt       = restic.SetupOptions{
+			ScratchDir:  restic.DefaultScratchDir,
+			EnableCache: false,
+		}
+		backupOpt = restic.BackupOptions{
+			StdinFileName: MongoDumpFile,
+		}
+		metrics = restic.MetricsOptions{
+			JobName: JobMongoBackup,
+		}
+	)
+
+	cmd := &cobra.Command{
+		Use:               "backup-mongo",
+		Short:             "Takes a backup of Mongo DB",
+		DisableAutoGenTag: true,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			flags.EnsureRequiredFlags(cmd, "app-binding", "provider", "secret-dir")
+
+			// prepare client
+			config, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfigPath)
+			if err != nil {
+				return err
+			}
+			kubeClient, err := kubernetes.NewForConfig(config)
+			if err != nil {
+				return err
+			}
+			appCatalogClient, err := appcatalog_cs.NewForConfig(config)
+			if err != nil {
+				return err
+			}
+
+			// get app binding
+			appBinding, err := appCatalogClient.AppcatalogV1alpha1().AppBindings(namespace).Get(appBindingName, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			// get secret
+			appBindingSecret, err := kubeClient.CoreV1().Secrets(namespace).Get(appBinding.Spec.Secret.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			// init restic wrapper
+			resticWrapper, err := restic.NewResticWrapper(setupOpt)
+			if err != nil {
+				return err
+			}
+			// hide password, don't print cmd
+			resticWrapper.HideCMD()
+
+			// setup pipe command
+			backupOpt.StdinPipeCommand = restic.Command{
+				Name: MongoDumpCMD,
+				Args: []interface{}{
+					"--host", appBinding.Spec.ClientConfig.Service.Name,
+					"--port", fmt.Sprint(appBinding.Spec.ClientConfig.Service.Port),
+					"--username", string(appBindingSecret.Data[MongoUser]),
+					"--password=" + string(appBindingSecret.Data[MongoPassword]),
+					"--archive",
```

Comment:
If it is possible, we should take dump and restore dump without `--archive`. If user needs archive, s/he can add it by `mongo-args`

Reply:
We need a single file for `restic stdin` backup. In `mongodump` `--archive` flag is used to backup data in a single file.

Reply:
xref: https://github.com/restic/restic/issues/2133